### PR TITLE
Implement Zone methods to allow open-ended zones (#1428)

### PR DESCRIPTION
**Pull request by tiqwab**
Implementation for https://github.com/scala-native/scala-native/issues/1428.

### Changes

- Added `open` to Zone companion object
- Added `close` and `isClosed` to Zone trait

### CODE REVIEW

**Comment by tiqwab** [comment]
CI failed due to stall? `test-runtime` succeeds in the local environment.

File: nativelib/src/main/scala/scala/scalanative/native/Zone.scala [hunk 1]
```diff
@@ -4,45 +4,57 @@ package native
...
 
   /** Allocates memory of given size. */
   def alloc(size: CSize): Ptr[Byte]
+
+  /** Frees allocations. This zone allocator is not reusable once closed. */
+  def close(): Unit
+
+  /** Return this zone allocator is closed or not. */
+  def isClosed: Boolean
+
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/native/Zone.scala]: Might add `def isOpen: Boolean = !isClosed` as well.

File: nativelib/src/main/scala/scala/scalanative/native/Zone.scala [hunk 2]
```diff
@@ -4,45 +4,57 @@ package native
...
+    private var node: Node      = null
+    private var closed: Boolean = false
+
+    final override def isClosed: Boolean = closed
 
     final def alloc(size: CSize): Ptr[Byte] = {
       val rawptr = libc.malloc(size)
       node = new Node(rawptr, node)
       fromRawPtr[Byte](rawptr)
     }
```
Comment (densh) [inline nativelib/src/main/scala/scala/scalanative/native/Zone.scala]: You should throw an exception if trying to alloc in a closed zone.

**Comment by densh** [comment]
Travis has been having hard time lately, it's likely not your fault. I've just restarted the CI.

**Comment by densh** [comment]
LGTM, thanks @tiqwab ! 👍

**Pull request merged**